flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerServiceImpl.java

```diff
@@ -117,6 +120,7 @@ public class InternalTimerServiceImpl<K, N> implements InternalTimerService<N> {
             startIdx = Math.min(keyGroupIdx, startIdx);
         }
         this.localKeyGroupRangeStartIdx = startIdx;
+        this.processingTimeCallback = this::onProcessingTime;
```

Comment:
It might be better to avoid introducing `processingTimeCallback`, or introduce it and make it a final variable. Given that the PR has already reused the name `advanceWatermark` in `InternalTimerServiceAsyncImpl` instead of using `asyncAdvanceWatermark`, it might be better to rename `onAsyncProcessingTime` back to `onProcessingTime` as well.

Reply:
The original idea was to avoid rewriting `startTimerService()` and `registerProcessingTimeTimer` by introducing `processingTimeCallback`, since it needs to be reassigned in the subclass, it cannot be marked as `final`.

For the `onAsyncProcessingTime`,  I changed it back to `onProcessingTime`, BTW, `onProcessingTime()` is a private method.